lib/safira/roulette/roulette.ex

```diff
@@ -196,4 +205,133 @@ defmodule Safira.Roulette do
   def change_attendee_prize(%AttendeePrize{} = attendee_prize) do
     AttendeePrize.changeset(attendee_prize, %{})
   end
+
+  @doc """
+  Transaction that take a number of tokens from an attendee,
+  apply a probability-based function for "spinning the wheel",
+  and give the price to the attendee.
+  """
+  def spin(attendee) do
+    Multi.new()
+    # remove tokens from attendee to spin the wheel
+    |> Multi.update(
+      :attendee,
+      Attendee.update_token_balance_changeset(attendee, %{
+        token_balance: attendee.token_balance - Application.fetch_env!(:safira, :roulette_cost)
+      })
+    )
+    # prize after spinning
+    |> Multi.run(:prize, fn _repo, _changes -> {:ok, spin_prize()} end)
+    # get an attendee_prize if exists one, or build one
+    |> Multi.run(:attendee_prize, fn _repo, %{attendee: attendee, prize: prize} ->
+      {:ok,
+       get_keys_attendee_prize(attendee.id, prize.id) ||
+         %AttendeePrize{attendee_id: attendee.id, prize_id: prize.id, quantity: 0}}
+    end)
```

Comment:
If the first Multi.run only yields the repo and the prize, where is the attendee used in line 226 coming from? Does the spin_prize function also return the attendee?

Reply:
Recall that Multi is a structure that holds the information for the transaction to run. That attendee in line 226 comes from the operation in the line 217. That attendee is the name of that operation. So what i'm doing is accessing the value of that operation that is stored in Multi structure. If the operation was called :op1, the line 226 would be %{op1: attendee,...}.